lib/matplotlib/legend.py

```diff
@@ -535,6 +536,21 @@ def val_or_rc(val, rc_name):
                     loc = locs[0] + ' ' + locs[1]
             # check that loc is in acceptable strings
             loc = _api.check_getitem(self.codes, loc=loc)
+        elif isinstance(loc, tuple):
+            # validate the tuple represents Real coordinates
+            if len(loc) != 2 or not (all(isinstance(e, numbers.Real)
+                                         for e in loc)):
+                raise ValueError(
+                    f"loc must be string or pair of numbers, not {loc!r}")
```

Comment:
Should this include a note about [0, 9] being allowed?

Reply:
Yes, but I'm unsure about the wording. Maybe "loc must be string, coordinate tuple or, an integer 0-10, not xyz"

Reply:
for bonus points you can also include the list of allowed strings!

One of the goals of our error messages should be, if it makes sense, to give the user enough of a hint about what they did wrong so that they can sort out how to fix it just from reading the traceback (and not have to go open a browser to search for the docs).

Reply:
This thought turned into #25292

Reply:
I updated the pull request to be more verbose. In your error message guidelines doc you mentioned that we should try to use the helper methods to standardize the error messages. I think that is good advice but it limits the ability to customize the error message to fit the context. To strike the balance in this situation, my suggestion is:

1. Use a less verbose error message when the user provides a non-string type such as " "loc must be string, coordinate tuple or, an integer 0-10, not 21"
2. Use the built in (and more verbose) error message when the user attempts to provide a string that does not match the allowed strings. This uses _api.check_getitem and yields a message like: "'xyz' is not a valid value for loc; supported values are 'best',  'upper right', 'upper left', 'lower left', 'lower right', 'right', 'center left', 'center right', 'lower center', 'upper center', 'center'"

Reply:
@tacaswell let me know if this tradeoff works for you or if you think I should update all the responses to include all the information. Either works for me. ^